packages/klighd-core/src/search-bar/searchbar-actions.ts

```diff
@@ -0,0 +1,680 @@
+    kind: typeof ToggleSearchBarAction.KIND
+    id: string
+    state?: 'show' | 'hide'
+    panel: SearchBarPanel
+}
+
+// eslint-disable-next-line no-redeclare
+export namespace ToggleSearchBarAction {
+    export const KIND = 'toggleSearchBar'
+
+    export function create(panel: SearchBarPanel, id: string, state?: 'show' | 'hide'): ToggleSearchBarAction {
+        return {
+            kind: KIND,
+            id,
+            state,
+            panel,
+        }
+    }
+
+    export function isThisAction(action: Action): action is ToggleSearchBarAction {
+        return action.kind === KIND
+    }
+}
+
+@injectable()
+export class ToggleSearchBarHandler implements IActionHandler {
+    handle(action: Action): void {
+        if (ToggleSearchBarAction.isThisAction(action)) {
+            if (action.id !== SearchBar.ID) return
+
+            const newVisible = action.state === 'show'
+
+            if (action.panel.isVisible !== newVisible) {
+                action.panel.changeVisibility(newVisible)
+                action.panel.update()
+            }
+        }
+    }
+}
+
+export interface UpdateHighlightsAction extends Action {
+    kind: typeof UpdateHighlightsAction.KIND
+    selectedIndex: number
+    previousIndex: number | undefined
+    results: SearchResult[]
+    panel: SearchBarPanel
+}
+
+// eslint-disable-next-line no-redeclare
+export namespace UpdateHighlightsAction {
+    export const KIND = 'updateHighlights'
+
+    export function create(
+        currentIndex: number,
+        prevIndex: number | undefined,
+        results: SearchResult[],
+        panel: SearchBarPanel
+    ): UpdateHighlightsAction {
+        return {
+            kind: KIND,
+            selectedIndex: currentIndex,
+            previousIndex: prevIndex,
+            results,
+            panel,
+        }
+    }
+
+    export function isThisAction(action: Action): action is UpdateHighlightsAction {
+        return action.kind === KIND
+    }
+}
+
+export interface ClearHighlightsAction extends Action {
+    kind: typeof ClearHighlightsAction.KIND
+}
+
+// eslint-disable-next-line no-redeclare
+export namespace ClearHighlightsAction {
+    export const KIND = 'clearHighlights'
+
+    export function create(): ClearHighlightsAction {
+        return {
+            kind: KIND,
+        }
+    }
+
+    export function isThisAction(action: Action): action is ClearHighlightsAction {
+        return action.kind === KIND
+    }
+}
+
+function createHighlightRectangle(
+    elem: SModelElement,
+    xPos: number,
+    yPos: number,
+    width: number,
+    height: number,
+    highlight: number
+): KRectangle {
+    return {
+        type: 'KRectangleImpl',
+        id: `highlightRect-${elem.id}`,
+        children: [],
+        properties: {
+            x: xPos,
+            y: yPos,
+            width,
+            height,
+            'klighd.rendering.highlight': highlight,
+        },
+        actions: [],
+        styles: [],
+    }
```

Comment:
In the long term, this should probably be part of a JS KGraph creation library. I would add a TODO notifying that here.